src/Avalonia.Base/Utilities/SmallDictionary.cs

```diff
@@ -77,7 +77,15 @@ void SetCore(TKey key, TValue value, bool overwrite)
         }
         else
         {
-            // We have a single element, upgrade to array
+            // We have a single element, check if we should update the value.
+            if((TKey)_data == key && overwrite)
```

Comment:
Nit: keep the `(TKey)_data` in a variable and reuse it below line 90 so we avoid a double cast.